extracted/block_07.py

```diff
@@ -0,0 +1,83 @@
+#Translate
+#Takes a string of DNA
+#Outputs dictionary of translated DNA sequence in all 6 reading frames
+#Keys should be f1, f2, f3, r1, r2, r3
+#Don't complement the sequence, just reverse 
+#Highlight stop codons with an *
+
+#Need Amino Acid dictionary to translate nucleotide codons
+#M is start, * is a stop
+
+aa_dict = { 
+    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', 
+    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', 
+    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', 
+    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  
+    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', 
+    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', 
+    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', 
+    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', 
+    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', 
+    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', 
+    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', 
+    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', 
+    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', 
+    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', 
+    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*', 
+    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W', 
+    } 
+
+
+def translate(sequence):
+    
+    #First translate sequence 
+    #sequence = sequence.replace("T", "U")
+    
+    # Make Upper case
+    sequence = sequence.upper()
+    
+    #Check sequence is DNA
+    # (Would swap T for U in the string below)
+    for i in sequence:
+        if i not in "ACGT":
+            raise BadSequenceException("Input string must be DNA Nucleotides only, no special characters")
+        
+    #Get length of sequence
+    length = len(sequence)
+        
+    #Define start position for reading frames
+    start_position = 0
+        
+    #Define reading frames, need to reduce the full length to avoid key errors
+    forward_1 = range(start_position, length - 2, 3)
```

Comment:
this could all be turned into a loop
```
for i in range(0, 3):
   forward = range(start_position + i, length - 2, 3)
   codons_f = ..
   codons_r =..
   # update dict
```